### PR TITLE
Fix and improve Target matching

### PR DESCRIPTION
**Pull request by mike-sul**
Signed-off-by: Mike Sul <mike.sul@foundries.io>

### CODE REVIEW

File: src/helpers.cc [hunk 1]
```diff
@@ -646,3 +646,24 @@ bool known_local_target(LiteClient& client, const Uptane::Target& t, std::vector
...
+  if (t1.length() != t2.length()) {
+    return false;
+  }
+
+  if (t1.filename() != t2.filename()) {
+    return false;
+  }
+
+  // match just OSTREE commit
+  if (t1.sha256Hash() != t2.sha256Hash()) {
```
Comment (doanac) [inline src/helpers.cc:664]: This is exactly how i'd always planned to implement this. I have wondered if making it generic like the upstream version does to account for sha512 might make sense. However, ostree can't do sha512, right? So I *think* the simplicity of this approach is better.

File: src/main.cc [hunk 2]
```diff
@@ -70,6 +70,15 @@ static std::unique_ptr<Uptane::Target> find_target(LiteClient& client,
...
   std::unique_ptr<Uptane::Target> latest = nullptr;
   for (auto& t : client.allTargets()) {
+
+    if (!t.IsValid()) {
+      continue;
+    }
+
+    if (!t.IsOstree()) {
+      continue;
+    }
```
Comment (doanac) [inline src/main.cc:80]: this is subtle, but I think you are trying to make things operate in a more secure manner? this makes this logic always return a valid target if there is one, right?

File: src/main.cc [hunk 3]
```diff
@@ -92,7 +101,6 @@ static std::unique_ptr<Uptane::Target> find_target(LiteClient& client,
 }
 
 static data::ResultCode::Numeric do_update(LiteClient& client, Uptane::Target target) {
-  target.InsertEcu({client.primary_ecu.first, client.primary_ecu.second});
```
Comment (doanac) [inline src/main.cc:95]: Have you tested that update events will still get reported to the server with this removed? Now that we have our own "TufClient" its probably fine, but in the old SotuptaneClient approach, this was required.

**Comment by doanac** [comment]
This looks fine. I think we can merge this. However, even though I think the 2nd commit looks good. I vote for updating the aktualizr recipe in meta-lmp with the the hash of the 1st commit for our v71 update. that way we are *only* pulling in one thing that we know fixes the one thing that was messing us up on Friday. Sound good?

File: src/helpers.cc (hunk 1 above)
Comment (mike-sul) [reply src/helpers.cc:664]: yes, just KISS

File: src/main.cc (hunk 2 above)
Comment (mike-sul) [reply src/main.cc:80]: yes, I am gonna improve here and yield some warning/error messages.

File: src/main.cc (hunk 3 above)
Comment (mike-sul) [reply src/main.cc:95]: I'll double check

**Comment by mike-sul** [comment]
> This looks fine. I think we can merge this. However, even though I think the 2nd commit looks good. I vote for updating the aktualizr recipe in meta-lmp with the the hash of the 1st commit for our v71 update. that way we are _only_ pulling in one thing that we know fixes the one thing that was messing us up on Friday. Sound good?

The second commit is required too because the first commit causes another issue - it fills ecus_ field  for the current target that is loaded from DB while Target from metadata doesn't have it so their comparison will always return false.

**Comment by mike-sul** [comment]
Also, I cannot properly verify it for the docker app. This fixes the issue that Ricardo ran into, I can check it as it happens before the issue I have :), but I cannot verify it from end-to-end perspective.

**Comment by mike-sul** [comment]
> I think this logic is only determining if an app has been added. The original logic also checks if an app has been removed

In general, I think we should get rid of docker|compose_app param in the config and let it be driven solely by Target so we have KISS :).

When are saying "added/removed" you mean in the config or on a file system?

**Comment by doanac** [comment]
we have compatibility to maintain, but I see my confusion now. `LiteClient::dockerAppsChanged()` is called at startup. I was thinking you were doing something with that. However, this is a different thing.  Now I'm trying to understand what's this is doing that targets_eq isn't.

**Comment by mike-sul** [comment]
> we have compatibility to maintain, but I see my confusion now. `LiteClient::dockerAppsChanged()` is called at startup. I was thinking you were doing something with that. However, this is a different thing. Now I'm trying to understand what's this is doing that targets_eq isn't.

So, it doesn't make sense to check if the app list in config has changed in the update cycle (isTargetCurrent, targets_eq) because aklite has to be restarted in order to apply config changes and if it's restarted then dockerAppsChanged() is called what would trigger an update if any change in the config?

**Comment by doanac** [comment]
> So, it doesn't make sense to check if the app list in config has changed in the update cycle (isTargetCurrent, targets_eq) because aklite has to be restarted in order to apply config changes and if it's restarted then dockerAppsChanged() is called what would trigger an update if any change in the config?

Yes. I think that's where we are getting mixed up here. I'm not sure we need the logic in `isTargetCurrent`. Its taking me some time to unpack and sort through mentally though.

**Comment by mike-sul** [comment]
> > So, it doesn't make sense to check if the app list in config has changed in the update cycle (isTargetCurrent, targets_eq) because aklite has to be restarted in order to apply config changes and if it's restarted then dockerAppsChanged() is called what would trigger an update if any change in the config?
> 
> Yes. I think that's where we are getting mixed up here. I'm not sure we need the logic in `isTargetCurrent`. Its taking me some time to unpack and sort through mentally though.

The only reason I put it there is for the case if the app(s) is removed somehow from a filesystem so isTargetCurrent() would detect it without a need in aklite restart, not sure if it makes sense to have it though.

**Comment by doanac** [comment]
> The only reason I put it there is for the case if the app(s) is removed somehow from a filesystem so isTargetCurrent() would detect it without a need in aklite restart, not sure if it makes sense to have it though. 

Its probably worth keeping around for compose-apps. Its one more way we keep users safe-r. However, i don't think its worth adding for docker-apps.

**Comment by mike-sul** [comment]
> Its probably worth keeping around for compose-apps. Its one more way we keep users safe-r. However, i don't think its worth adding for docker-apps.

ok, agree. I finished testing and see that I need to add dockerAppsChanged() for the non-daemon mode, the rest looks ok.

**Comment by doanac** [comment]
one thing we have missing - i guess we can do it in this PR - is to remove the "force" option for `aktualir-lite update`. We should just force in that mode and assume its a "expert" type decision

**Comment by mike-sul** [comment]
@doanac It should work as we expect now.

**Pull request merged**